joblib/numpy_pickle_utils.py

```diff
@@ -0,0 +1,199 @@
+"""Utilities for fast persistence of big data, with optional compression."""
+
+# Author: Gael Varoquaux <gael dot varoquaux at normalesup dot org>
+# Copyright (c) 2009 Gael Varoquaux
+# License: BSD Style, 3 clauses.
+
+import pickle
+import sys
+import io
+
+PY3 = sys.version_info[0] >= 3
+PY26 = sys.version_info[0] == 2 and sys.version_info[1] == 6
+
+if PY3:
+    Unpickler = pickle._Unpickler
+    Pickler = pickle._Pickler
+    xrange = range
+else:
+    Unpickler = pickle.Unpickler
+    Pickler = pickle.Pickler
+
+
+def hex_str(an_int):
+    """Convert an int to an hexadecimal string."""
+    return '{0:#x}'.format(an_int)
+
+if PY3:
+    def asbytes(s):
+        if isinstance(s, bytes):
+            return s
+        return s.encode('latin1')
+else:
+    asbytes = str
+
+try:
+    import numpy as np
+except ImportError:
+    np = None
+
+# Compressed pickle header format: _ZFILE_PREFIX followed by
+# bytes which contains the length of the zlib compressed data as an
+# hexadecimal string. For example: 'ZF0x139              '
+_ZFILE_PREFIX = asbytes('ZF')
+_GZIP_PREFIX = b'\x1f\x8b'
+
+
+def gzip_file_factory(f, mode='rb', compresslevel=0):
```

Comment:
> I would have the mode and compresslevel as part of the __init__ signature to mirror gzip.GzipFile.

Scratch that, you haven't actually overriden the `__init__` I got confused with _init_write ...